### PR TITLE
[Merged by Bors] - feat(Algebra/Homology): behaviour of the total complex with respect to the shifts

### PR DESCRIPTION
**Pull request by joelriou**
---
- [x] depends on: #10764
- [x] depends on: #10711

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Homology/HomologicalBicomplex.lean [hunk 1]
```diff
@@ -203,4 +203,16 @@ def flipEquivalence :
...
+variable (K : HomologicalComplex₂ C c₁ c₂)
+
+/-- The obvious isomorphism `(K.X x₁).X x₂ ≅ (K.X y₁).X y₂` when `x₁ = y₁` and `x₂ = y₂`. -/
+def XXIsoOfEq {x₁ y₁ : I₁} (h₁ : x₁ = y₁) {x₂ y₂ : I₂} (h₂ : x₂ = y₂) :
+    (K.X x₁).X x₂ ≅ (K.X y₁).X y₂ :=
+  eqToIso (by subst h₁ h₂; rfl)
+
+@[simp]
+lemma XXIsoOfEq_rfl (i₁ : I₁) (i₂ : I₂) :
+    K.XXIsoOfEq (rfl : i₁ = i₁) (rfl : i₂= i₂) = Iso.refl _ := rfl
```
Comment (TwoFX) [inline Mathlib/Algebra/Homology/HomologicalBicomplex.lean:215]:
```suggestion
    K.XXIsoOfEq (rfl : i₁ = i₁) (rfl : i₂ = i₂) = Iso.refl _ := rfl
```

File: Mathlib/Algebra/Homology/TotalComplexShift.lean [hunk 2]
```diff
@@ -0,0 +1,206 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.Algebra.Homology.HomotopyCategory.Shift
+import Mathlib.Algebra.Homology.TotalComplex
```
Comment (TwoFX) [inline Mathlib/Algebra/Homology/TotalComplexShift.lean:7]:
```suggestion
import Mathlib.Algebra.Homology.TotalComplex

```

File: Mathlib/Algebra/Homology/TotalComplexShift.lean [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    (K.XXIsoOfEq rfl (Int.sub_add_cancel q y)).inv ≫
+      ((shiftFunctor₂ C y).obj K).ιTotal (up ℤ) p (q - y) n (by dsimp at hpq ⊢; omega))
+  hom_inv_id := by
+    ext p q h
+    dsimp
+    simp only [ι_totalDesc_assoc, Linear.units_smul_comp, ι_totalDesc, smul_smul,
+      Int.units_mul_self, one_smul, comp_id]
+    exact ((shiftFunctor₂ C y).obj K).XXIsoOfEq_inv_ιTotal _ rfl (by omega) _ _
+
+lemma D₁_totalShift₂XIso_hom (n₀ n₁ n₀' n₁' : ℤ) (h₀ : n₀ + y = n₀') (h₁ : n₁ + y = n₁') :
```
Comment (TwoFX) [inline Mathlib/Algebra/Homology/TotalComplexShift.lean:148]: Should this get `@[reassoc]`? Same for `D₂_totalShift₂XIso_hom` below and the two analogous lemmas above.

Comment (joelriou) [reply Mathlib/Algebra/Homology/TotalComplexShift.lean:150]: Thanks! Yes, I have just added the reassoc attribute.

File: Mathlib/Algebra/Homology/TotalComplexShift.lean [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    (K.XXIsoOfEq (Int.sub_add_cancel p x) rfl).inv ≫
+      ((shiftFunctor₁ C x).obj K).ιTotal (up ℤ) (p - x) q n
+        (by dsimp at hpq ⊢; omega))
+  hom_inv_id := by
+    ext p q h
+    dsimp
+    simp only [ι_totalDesc_assoc, CochainComplex.shiftFunctor_obj_X', ι_totalDesc, comp_id]
+    exact ((shiftFunctor₁ C x).obj K).XXIsoOfEq_inv_ιTotal _ (by omega) rfl _ _
+
+lemma D₁_totalShift₁XIso_hom (n₀ n₁ n₀' n₁' : ℤ) (h₀ : n₀ + x = n₀') (h₁ : n₁ + x = n₁') :
```
Comment (TwoFX) [inline Mathlib/Algebra/Homology/TotalComplexShift.lean:76]: I think this one and the lemma below are also missing `@[reassoc]`?

**Review by TwoFX — APPROVED** [review]
Thanks!
bors d+

File: Mathlib/Algebra/Homology/TotalComplexShift.lean (hunk 4 above)
Comment (joelriou) [reply Mathlib/Algebra/Homology/TotalComplexShift.lean:77]: Thanks for spotting this!

**Comment by joelriou** [comment]
Thanks for the review!

bors merge

**Pull request closed without merging**